homework 1 and 2

VitaliyBogatov opened the pull request.

Commit:
homework 1 and 2

numash commented on TI_Practice1/Circle.cs:
```diff
@@ -0,0 +1,28 @@
+            radius = val;
+        }
+
+        readonly double radius;
```
1. Could you please name the class field starting with "_" - _radius?
2. Could you also please add a property for the _radius field?
3. In addition, the class components should be arranged in the following order:
- fields
- properties
- methods

VitaliyBogatov replied:
1. DONE. But can you please explain why should I do it (probably during the next practice or lecture)? As I understand we should use the underscore in order not to create new variables' names when this. is not used. Isn't it?
2. DONE
3. DONE
4. DONE. As I understood readonly is an access modifier too but it can be used with properties?

numash replied:
sure, let's discuss it if you still have questions

numash commented on TI_Practice1/Circle.cs:
```diff
@@ -0,0 +1,28 @@
+        public double GetArea()
+        {
+            resp = Math.PI * Math.Pow(radius, 2);
+            return resp;
```
The "resp" variable is not needed here, you can simply use
return Math.PI * Math.Pow(radius, 2);

It is better to get rid of all the unnecessary variables.

Furthermore, in this particular case, you're using a class field. This is not correct because a class field represents some semantic property. E.g. class Car will have engine, wheels, etc. fields.
All the other subsidiary values should be in the local variables.

VitaliyBogatov replied:
DONE. Can we discuss class fields and properties usage during the next practice/lecture

numash commented on TI_Practice1/Practice1.cs:
```diff
@@ -0,0 +1,48 @@
+            else
+            {
+                Console.Write("\nCircle can be put inside square.");
+            }
```
I believe this solution is not correct.
It says "Circle can be put inside square." for the circle radius = 4 and square side = 6.

![image](https://user-images.githubusercontent.com/9382934/109701636-e6f69880-7b9b-11eb-84c2-8131315b56c2.png)

Could you please review the solution?

VitaliyBogatov replied:
DONE. My faill

VitaliyBogatov replied:
DONE My fail

numash commented on TI_Practice1/Practice1.cs:
```diff
@@ -0,0 +1,48 @@
+                Console.Write("\nCircle can be put inside square.");
+            }
+
+            Console.WriteLine("\n\nPlease eneter any button to finish....");
```
You can't actually enter any button as far as I know :)
"Press any button" maybe?

VitaliyBogatov replied:
DONE )

numash commented on TI_Practice1/Circle.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Practice1
+{
+    class Circle
```
could you please specify the access modifier for the class? it is better to set it explicitly

VitaliyBogatov replied:
public access modifier has added

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+            while (matchCount <= _attempts)
+            {
+                var input = Console.ReadLine();
+                input = input.Replace(',', '.');
```
I can't enter any decimal value for now

![image](https://user-images.githubusercontent.com/9382934/109702609-1fe33d00-7b9d-11eb-959c-5b4720a1444b.png)

VitaliyBogatov replied:
Not able to reproduce using EN/RU keyboards http://prntscr.com/10bmxsg

numash replied:
it depends on the locale of the machine.
Actually you can fix this issue = make your program understand all the locales, but I suggest doing it only if you have time in future.

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+            {
+                var input = Console.ReadLine();
+                input = input.Replace(',', '.');
+                //String pattern = @"^[1-9]\d*([\.|\,]\d+)?$$";
```
We don't usually leave comments in code if they are not absolutely needed

VitaliyBogatov replied:
it is just another implementation because I do not understand the practice done criteria
So I decided to provide other option in the commented code

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+                if (double.TryParse(input, out val) && val >0)
+                {
+                    val = Math.Round(val, 2);
+                    break;
```
why do you need break here?

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+                {
+                    if (matchCount < _attempts)
+                    {
+                        Console.Write($"Only numbers greater than 0, comma or dot separators are allowed.\nPlese reenter a number. {(_attempts - matchCount)} attempt(s) left: ");
```
typo - "Please"

VitaliyBogatov replied:
DONE

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+                    }
+                    else if (matchCount == 3)
+                    {
+                        var random = new Random();
```
this part would be better in a separate method

VitaliyBogatov replied:
DONE

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+                    else if (matchCount == 3)
+                    {
+                        var random = new Random();
+                        const double min = 0.5;
```
constants should also be declared above the methods

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+            _attempts = attempts;
+        }
+
+        public double IsNumber()
```
method name that starts with "Is.." should return bool value.

I suggest you name it something like GetValueFromConsole() or anything else that reflects the actions you do inside

VitaliyBogatov replied:
DONE

numash commented on TI_Practice1/Validator.cs:
```diff
@@ -0,0 +1,54 @@
+                        Console.Write("\nValue is set on default to " + val);
+                    }
+
+                    matchCount++;
```
you can avoid this action if you use while (matchCount++ <= _attempts)

numash replied:
in general, it looks like this if-else-if-else part works, but it is very complicated and hard to read. I suggest you simplify it.
Please, ask trainers if you don't have an idea

VitaliyBogatov replied:
1. Cannot use  if (matchCount++ < _ attempts) because matchCount increased just after the if condition as result {(_attempts - matchCount)} will be 3-2 = 1 attempt

Commit:
With fixes

Commit:
Fixed

Commit:
Refactored

numash commented:
1. All classes should have access modifiers (Validator, Practice)
2. In this particular case with _radius and _side you can use the auto property. Please, read about it and ask if anything’s unclear.
3. GetArea() (and other similar methods) can in such cases be a property: public double Area => Math.Pow(_side, 2);
4. Please, remove the unused dependencies.
5. Could you please get rid of the «result» variable in the Main method?
6. Could you please move the costs in the Validator class above the fields and name it using PascalCase?
7. Have you already got all the answers you needed? Or is there anything unclear?

A piece of advice:
I would make the _attempts filed into constant of the Validator class (or even better - in a separate file) and make the Validator static class. The reason for that is that object creation for this class seems to be not needed from the perspective of the meaning of this class. So in that case you could just write Validator.GetValueFromConsole() in Main.

Commit:
Issues have been fixed
1. access modifiers are added to all classes
2. auto properties are added
3. GetArea(), GetDiagonal(), GetDiametr() methods have been modified to properties because they have one action only
4. unused dependencies have been removed

Commit:
fIssues FIxed

Commit:
Issues fixed

Commit:
Properties have been fixed
